director/projects/api/views/nodes.py

```diff
@@ -0,0 +1,222 @@
+        return [permission() for permission in permission_classes]
+
+    @swagger_auto_schema(
+        request_body=NodesCreateRequest,
+        responses={status.HTTP_201_CREATED: NodesCreateResponse},
+    )
+    def create(self, request: Request) -> Response:
+        """
+        Create a node.
+
+        Receives a request with the `node` and other information e.g. `project`.
+        Returns the URL of the node.
+        """
+        serializer = NodesCreateRequest(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        project = serializer.validated_data.get("project")
+        app = serializer.validated_data.get("app")
+        host = serializer.validated_data.get("host")
+        node = serializer.validated_data.get("node")
+
+        # Check that the user has EDIT permissions for the project
+        if not self.is_permitted(
+            request.user, ProjectPermissionType.EDIT, pk=project.id
+        ):
+            raise PermissionDenied
+
+        fingerprint = "{}-{}-{}-{}".format(project, app, host, json.dumps(node))
+        key = hashlib.sha256(fingerprint.encode("utf-8")).hexdigest()
+
+        # Get (efficiently, if it exists), or create, the node
+        try:
+            node = Node.objects.get(key=key)
+        except Node.DoesNotExist:
+            node = Node.objects.create(
+                key=key, project=project, app=app, host=host, json=node
+            )
+
+        serializer = NodesCreateResponse(node, context={"request": request})
+        return Response(serializer.data, status=status.HTTP_201_CREATED,)
+
+    @swagger_auto_schema(responses={status.HTTP_200_OK: NodeSerializer},)
+    def retrieve(
+        self, request: Request, key: str, format: Optional[str] = None
+    ) -> Response:
+        """
+        Retrieve a node.
+
+        Performs content negotiation and authorization based on the `Accept` header.
+
+        - For `application/json` returns 403 if the request user is not authorized
+        to read the project.
+        - For `text/html` (and others) returns a very simple HTML rendering of the
+        of the node (only type etc) if the user does not have read access to the project.
+        Otherwise, should return a full HTML rendering of the node using Encoda.
+        """
+        node = get_object_or_404(Node, key=key)
+        accept = request.META.get("HTTP_ACCEPT")
+        if format == "json" or (accept and "application/json" in accept):
+            # Require the user is authenticated and has VIEW permissions for the project
+            if not request.user.is_authenticated:
+                raise NotAuthenticated
+            if not self.is_permitted(
+                request.user, ProjectPermissionType.VIEW, pk=node.project.id
+            ):
+                raise PermissionDenied
+
+            serializer = NodeSerializer(node, context={"request": request})
+            return Response(serializer.data)
+        else:
+            if self.is_permitted(
```

Comment:
Previously the check was `if not self.is_permitted` with an exception for short circuit exit. I suggest following the same logic, if the user is not permitted here, return the basic Response immediately.

Reply:
Done.